### PR TITLE
Creates target module and moves ClassBalance to it

### LINKED ISSUE #334: Create target package for supervised target visualizers
For supervised machine learning, there is some independent analysis that must be conducted on the target vector; for example label encoding, "bucketization", etc. 

For visualizers:

- move the class balance here without requiring a model, fit, or score 
- perform detection of sequential or categorical data (and associated colors) 
- plot feature/target correlations in this package

### PR DESCRIPTION
**Pull request by bbengfort**
The target module allows independent analysis to be conducted on the dependent variable in supervised machine learning. For example, analysis of label encoding, binarization, or thresholding mechanisms, to simpler analyses like class balance or feature correlation. The base class, `TargetVisualizer`, implements a fit method that only accepts a vector y, similar to the `LabelEncoder` transformer in scikit-learn.

By moving the `ClassBalanc`e visualizer to this package, it no longer requires a model (and is not a score visualizer) so the class balance analysis can be done a priori without requiring a model. This seems more in line with many workflows that create train/test splits then would like to know the class balance of those splits before moving on to evaluation methods.

`ClassBalance` has also been extended to provide two modes. In "balance" mode, the default mode, the bar graph is simply drawn for each class. In "compare" mode, the train and test splits for each class are drawn side by side.

Here are the examples of the two new class balance visualizations:

class_balance

class_balance_compare

Fixes #334

### CODE REVIEW

**Comment by bbengfort** [comment]
@DistrictDataLabs/team-oz-maintainers this is ready for review -- I realize it's quite a significant change, so I went through and added some comments. I'd appreciate it if a couple of you could take a look at this for discussion!

File: docs/api/classifier/class_balance.py [hunk 1]
```diff
@@ -1,29 +0,0 @@
-import pandas as pd
```
Comment (bbengfort) [inline docs/api/classifier/class_balance.py]: Moved to `docs/api/target/class_balance.py`

File: docs/api/classifier/class_balance.rst [hunk 2]
```diff
@@ -1,51 +0,0 @@
-.. -*- mode: rst -*-
```
Comment (bbengfort) [inline docs/api/classifier/class_balance.rst]: Moved to `docs/api/target/class_balance.rst`

File: docs/api/classifier/class_prediction_error.rst [hunk 3]
```diff
@@ -45,7 +45,7 @@ The class prediction error chart provides a way to quickly understand how good y
 API Reference
 -------------
 
-.. automodule:: yellowbrick.classifier.class_balance
+.. automodule:: yellowbrick.classifier.class_prediction_error
```
Comment (bbengfort) [inline docs/api/classifier/class_prediction_error.rst]: Renamed the module with `ClassPredictionError` and separated it from `ClassBalance`

File: docs/api/classifier/index.rst [hunk 4]
```diff
@@ -8,7 +8,7 @@ Classification models attempt to predict a target in a discrete space, that is a
 -  :doc:`classification_report`: A visual classification report that displays precision, recall, and F1 per-class as a heatmap.
 -  :doc:`confusion_matrix`: A heatmap view of the confusion matrix of pairs of classes in multi-class classification.
 -  :doc:`rocauc`: Graphs the receiver operating characteristics and area under the curve.
--  :doc:`class_balance`: Visual inspection of the target to show the support of each class to the final estimator.
+-  :doc:`../target/class_balance`: Visual inspection of the target to show the support of each class to the final estimator.
```
Comment (bbengfort) [inline docs/api/classifier/index.rst]: Am still leaving the `ClassBalance` visualizer with the classification documentation because I know folks will want to reference it from there, and also because there isn't much documentation for the target module yet.

File: docs/gallery.rst [hunk 5]
```diff
@@ -1,10 +1,7 @@
 .. -*- mode: rst -*-
 
-====================
-Example Gallery
-=======
 Gallery
-====================
+=======
```
Comment (bbengfort) [inline docs/gallery.rst]: Not sure why this header was like it, but Sphinx was complaining so I updated it.

File: docs/gallery.rst [hunk 6]
```diff
@@ -125,11 +122,11 @@ Clasification Visualizers
...
     :target: api/classifier/rocauc.html#rocauc
 
-.. image:: api/classifier/images/class_balance.png
+.. image:: api/target/images/class_balance_compare.png
     :width: 200px
     :height: 100px
-    :alt: Class Balance for Random Forest Classifier
-    :target: api/classifier/class_balance.html#class-balance
+    :alt: Class Balance
+    :target: api/target/class_balance.html#class-balance
```
Comment (bbengfort) [inline docs/gallery.rst]: Should eventually also add the regular class balance to the gallery, but left the compare classifier since that is what was there before.

File: tests/test_classifier/test_class_balance.py [hunk 7]
```diff
@@ -1,61 +0,0 @@
-
```
Comment (bbengfort) [inline tests/test_classifier/test_class_balance.py]: Moved to `tests/test_target/test_class_balance.py`

File: tests/test_classifier/test_class_prediction_error.py [hunk 8]
```diff
@@ -65,7 +65,7 @@ def test_class_prediction_error_quickmethod(self):
         ax = fig.add_subplot()
 
         clf = LinearSVC(random_state=42)
-        g = class_prediction_error(clf, X, y, ax)
+        g = class_prediction_error(clf, X, y, ax, random_state=42)
```
Comment (bbengfort) [inline tests/test_classifier/test_class_prediction_error.py]: Gave the quick method a random state argument and removed the random state hard coding in that function.

File: yellowbrick/classifier/__init__.py [hunk 9]
```diff
@@ -23,8 +23,11 @@
...
 from .base import ClassificationScoreVisualizer
-from .class_balance import ClassBalance, ClassPredictionError
+from .class_prediction_error import ClassPredictionError, class_prediction_error
 from .classification_report import ClassificationReport, classification_report
 from .confusion_matrix import ConfusionMatrix, confusion_matrix
 from .rocauc import ROCAUC, roc_auc
 from .threshold import DiscriminationThreshold, discrimination_threshold
+
+## Import from target for backward compatibility and classifier association
+from ..target.class_balance import ClassBalance, class_balance
```
Comment (bbengfort) [inline yellowbrick/classifier/__init__.py]: Still importing class balance into the classification module for easy reference (though it does break backward compatibility)

File: yellowbrick/classifier/class_prediction_error.py [hunk 10]
```diff
@@ -1,194 +1,33 @@
...
 
 from ..exceptions import ModelError, YellowbrickValueError
 from ..style.colors import resolve_colors
 
 
-##########################################################################
-## Class Balance Chart
-##########################################################################
-
-class ClassBalance(ClassificationScoreVisualizer):
```
Comment (bbengfort) [inline yellowbrick/classifier/class_prediction_error.py]: Most of the changes here are removing `ClassBalance` and keeping `ClassPredictionError`

File: yellowbrick/classifier/class_prediction_error.py [hunk 11]
```diff
@@ -367,9 +226,9 @@ def class_prediction_error(model, X, y=None, ax=None, classes=None,
     visualizer = ClassPredictionError(model, ax, classes, **kwargs)
 
     # Create the train and test splits
-    X_train, X_test, y_train, y_test = train_test_split(X, y,
-                                                        test_size=test_size,
-                                                        random_state=42)
```
Comment (bbengfort) [inline yellowbrick/classifier/class_prediction_error.py]: Removed the hardcoded random state

File: yellowbrick/datasets/download.py [hunk 12]
```diff
@@ -113,8 +116,9 @@ def load_game(data_path=FIXTURES):
     """
     # name of the dataset
     name = 'game'
-    data = load_numpy(name, data_path=data_path)
-    return data
+    path = _lookup_path(name, data_path=data_path)
+    dtype = np.array(['S1']*42+['|S4'])
+    return np.genfromtxt(path, dtype=dtype, delimiter=',', names=True)
```
Comment (bbengfort) [inline yellowbrick/datasets/download.py]: Had to make these changes to get the documentation example to work. Changes still required in our load datasets methodology.

File: docs/gallery.rst [hunk 13]
```diff
@@ -104,7 +101,7 @@ Regression Visualizers
     :target: api/regressor/alphas.html#alpha-selection
 
 
-Clasification Visualizers
+Classification Visualizers
```
Comment (ndanielsen) [inline docs/gallery.rst]: good catch

File: docs/gallery.rst [hunk 14]
```diff
@@ -194,7 +191,7 @@ Text Modeling Visualizers
     :alt: TSNE Projection of Documents
     :target: api/text/tsne.html#t-sne-corpus-visualization
 
-Decision Boundaries Vizualizer
+Decision Boundaries Visualizer
```
Comment (ndanielsen) [inline docs/gallery.rst]: damn, that must have been my typo

File: yellowbrick/classifier/__init__.py (hunk 9 above)
Comment (ndanielsen) [reply yellowbrick/classifier/__init__.py]: do we want to add a deprecation warning here or somewhere else?

File: yellowbrick/datasets/download.py (hunk 12 above)
Comment (ndanielsen) [reply yellowbrick/datasets/download.py]: yeah, i noticed that earlier that some of the dtypes still need to be defined properly

File: yellowbrick/classifier/__init__.py (hunk 9 above)
Comment (bbengfort) [reply yellowbrick/classifier/__init__.py]: I'm not actually sure how to add the deprecation warning - the deprecated part is that the function signature has changed, but the function is still there. I suppose I could check the type of the input and raise an exception; I'll do that in a follow on pull request so it's easier to see. 

As for deprecating this import - I'm also not exactly sure how to do that; but I do like this particular visualizer living in "two" places - target and classifier; since it's a classification based target analysis. What do you think?
